### PR TITLE
Split out quic- and udp-client definitions

### PR DESCRIPTION
**Pull request by CriesofCarrots**
#### Problem
Currently, it's impossible to import just a udp-client *or* a quic-client; a user must pull both, including all their constituent dependencies.

#### Summary of Changes
- (Groundwork) Move ConnectionCache, TpuConnection, TpuClient, and ThinClient back into `solana-client` (deprecation to come in a separate PR). This part feels like a step backward, since these structs came from `solana-client` originally. But it is the best way I've found to preserve solana-client backward compatibility and allow stuff to be stripped out of `solana-tpu-client` so it can be lower than `solana-quic-client` and `solana-udp-client` crates. This will all feel much nicer when the whole repo is transitioned to TpuConnectionCache.
- Transition the new TpuClient and ThinClient structs to use TpuConnectionCache
- Split out `quic-client` and `udp-client` to their respective crates 🥳 

Toward #28401 

Please review by commit to retain your sanity. (First 4 commits don't compile.) "Hide whitespace" also recommended.

### CODE REVIEW

File: tpu-client/src/nonblocking/tpu_client.rs [hunk 1]
```diff
@@ -46,201 +46,231 @@ use {
...
-    #[error("RPC error: {0:?}")]
-    RpcError(#[from] ClientError),
-    #[error("IO error: {0:?}")]
-    IoError(#[from] std::io::Error),
-    #[error("Signer error: {0:?}")]
-    SignerError(#[from] SignerError),
-    #[error("Custom error: {0}")]
-    Custom(String),
-}
+pub mod temporary_pub {
```
Comment (CriesofCarrots) [inline tpu-client/src/nonblocking/tpu_client.rs:49]: In order to deduplicate code between solana-tpu-client and solana-client, somethings needed to be made public. I wrapped them in these `temporary_pub` modules to make them easier to unpub in the future.
Thoughts on this approach very welcome.

Comment (ryleung-solana) [reply tpu-client/src/nonblocking/tpu_client.rs:49]: What's preventing us from just moving all the code into the solana-client module?

Comment (CriesofCarrots) [reply tpu-client/src/nonblocking/tpu_client.rs:49]: If by "all the code" you mean the code in these `temporary_pub` modules, what's preventing us is that the code is being used in the lower-level crates as well.

**Review by t-nelson — APPROVED** [review]
lgtm. thanks for the excellent per-commit breakdown!

thoughts on moving the new crates under the `client` dir in the monorepo, ala. `sdk/program`, for the sake of organization? i think i'm mostly indifferent, but _am_ starting to get annoyed at the number of dirs at top-level

**Comment by KirillLykov** [comment]
I looked through the PR commit by commit and on the high level I don't see any pitfalls. General idea is good and I hope this work will lead to the simplification of clients code. But since I don't know all the logic behind TpuClient, it is hard to say for sure that each change doesn't bring any harm.

**Comment by CriesofCarrots** [comment]
> thoughts on moving the new crates under the `client` dir in the monorepo

Thanks, I like it. We are reaching the point where an extra layer of organization would be nice. I'll do that in a follow-up

**Comment by CriesofCarrots** [comment]
Client build failure unrelated

**Pull request merged**